src/rust/engine/process_execution/src/remote.rs

```diff
@@ -890,6 +687,208 @@ fn make_execute_request(
   Ok((action, command, execute_request))
 }
 
+pub fn populate_fallible_execution_result(
+  store: Store,
+  execute_response: bazel_protos::remote_execution::ExecuteResponse,
+  execution_attempts: Vec<ExecutionStats>,
+) -> impl Future<Item = FallibleExecuteProcessResult, Error = String> {
+  extract_stdout(&store, &execute_response)
+    .join(extract_stderr(&store, &execute_response))
+    .join(extract_output_files(store, &execute_response))
+    .and_then(move |((stdout, stderr), output_directory)| {
+      Ok(FallibleExecuteProcessResult {
+        stdout: stdout,
+        stderr: stderr,
+        exit_code: execute_response.get_result().get_exit_code(),
+        output_directory: output_directory,
+        execution_attempts: execution_attempts,
+      })
+    })
+}
+
+fn extract_stdout(
+  store: &Store,
+  execute_response: &bazel_protos::remote_execution::ExecuteResponse,
+) -> BoxFuture<Bytes, String> {
+  if execute_response.get_result().has_stdout_digest() {
+    let stdout_digest_result: Result<Digest, String> =
+      execute_response.get_result().get_stdout_digest().into();
+    let stdout_digest =
+      try_future!(stdout_digest_result.map_err(|err| format!("Error extracting stdout: {}", err)));
+    store
+      .load_file_bytes_with(stdout_digest, |v| v)
+      .map_err(move |error| {
+        format!(
+          "Error fetching stdout digest ({:?}): {:?}",
+          stdout_digest, error
+        )
+      })
+      .and_then(move |maybe_value| {
```

Comment:
Should this method and branch return an `Option` instead so that this case can be handled differently by the caller for the cache vs other cases? Ditto the other methods.

Reply:
Can do, but the only difference I can really imagine is logging verbosity; I would expect `None` to not-or-barely log, but `Err` to log saying there was a communication error. We can happily do this as a follow-up if we decide to change the caller's behaviour :)

Reply:
I was thinking that it would avoid creating error strings for cache misses. But yea, minor.